governance/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+}
+
+#[derive(TypeAbi, TopDecode)]
+pub struct CallData<M: ManagedTypeApi> {
+    pub endpoint_name: ManagedBuffer<M>,
+    pub arguments: ManagedVec<M, ManagedBuffer<M>>,
+}
+
+#[derive(TypeAbi, TopDecode)]
+pub struct ExecutePayload<M: ManagedTypeApi> {
+    pub command: GovernanceCommand,
+    pub target: ManagedAddress<M>,
+    pub call_data: ManagedBuffer<M>,
+    pub native_value: BigUint<M>,
+    pub eta: u64,
+}
+
+#[multiversx_sc::contract]
+pub trait Governance: events::Events {
+    #[init]
+    fn init(
+        &self,
+        gateway: ManagedAddress,
+        governance_chain: ManagedBuffer,
+        governance_address: ManagedBuffer,
+        minimum_time_delay: u64,
+    ) {
+        require!(
+            !gateway.is_zero() && !governance_chain.is_empty() && !governance_address.is_empty(),
+            "Invalid address"
+        );
+
+        self.gateway().set_if_empty(gateway);
+        self.minimum_time_lock_delay()
+            .set_if_empty(minimum_time_delay);
+
+        self.governance_chain().set_if_empty(&governance_chain);
+        self.governance_address().set_if_empty(&governance_address);
+        self.governance_chain_hash()
+            .set_if_empty(self.crypto().keccak256(&governance_chain));
+        self.governance_address_hash()
+            .set_if_empty(self.crypto().keccak256(&governance_address));
+    }
+
+    #[upgrade]
+    fn upgrade(&self) {}
+
+    #[payable("EGLD")]
+    #[endpoint(executeProposal)]
+    fn execute_proposal(
+        &self,
+        target: ManagedAddress,
+        call_data: ManagedBuffer,
+        native_value: BigUint,
+    ) {
+        let proposal_hash = self.get_proposal_hash(&target, &call_data, &native_value);
+
+        let eta = self.finalize_time_lock(&proposal_hash);
+
+        self.proposal_executed_event(
+            &proposal_hash,
+            &target,
+            ProposalEventData {
+                call_data: &call_data,
+                value: &native_value,
+            },
+        );
+
+        let call_data: CallData<Self::Api> = CallData::<Self::Api>::top_decode(call_data)
+            .unwrap_or_else(|_| sc_panic!("Could not decode call data"));
+
+        self.send()
+            .contract_call::<()>(target, call_data.endpoint_name)
+            .with_egld_transfer(native_value)
+            .with_raw_arguments(call_data.arguments.into())
+            .async_call()
+            .with_callback(
+                self.callbacks()
+                    .execute_proposal_callback(&proposal_hash, eta),
+            )
+            .call_and_exit();
+    }
+
+    // Can only be called by self (through the execute_proposal endpoint)
+    #[endpoint(withdraw)]
+    fn withdraw(&self, recipient: ManagedAddress, amount: BigUint) {
+        require!(
+            self.blockchain().get_caller() == self.blockchain().get_sc_address(),
+            "Not self"
+        );
+
+        self.send().direct_egld(&recipient, &amount);
+    }
+
+    #[endpoint]
+    fn execute(
```

Comment:
A little bit too generic the naming. Maybe execute_command, as you call the process_command function at the end?

Reply:
This needs to be called like this since `execute` endpoint is the convention used for an endpoint that can be executed by Axelar protocol, similar to Solidity implementation:
https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/main/contracts/executable/AxelarExecutable.sol#L17